packages/embark-deployment/src/index.js

```diff
@@ -70,20 +70,32 @@ class DeployManager {
                 if (typeof result === 'function') {
                   callback = result;
                 }
-                contract._gasLimit = self.gasLimit;
-                contract._skipGasEstimations = self.skipGasEstimations;
-                self.events.request('deploy:contract', contract, (err) => {
-                  if (err) {
-                    contract.error = err.message || err;
-                    if (contract.error === constants.blockchain.gasAllowanceError) {
-                      self.logger.error(`[${contract.className}]: ${constants.blockchain.gasAllowanceErrorMessage}`);
+                if (contract.addressHandler && typeof contract.addressHandler === 'function') {
+                  self.plugins.runActionsForEvent('contract:address:handler', {contract}, (err, address) => {
+                    if (err) {
+                      errors.push(err);
                     } else {
-                      self.logger.error(`[${contract.className}]: ${err.message || err}`);
+                      contract.address = address;
+                      self.logger.info(contract.className.bold.cyan + __(' already deployed at ').green + contract.address.bold.cyan);
                     }
-                    errors.push(err);
-                  }
-                  callback();
-                });
+                    callback();
+                  });
+                } else {
+                  contract._gasLimit = self.gasLimit;
+                  contract._skipGasEstimations = self.skipGasEstimations;
+                  self.events.request('deploy:contract', contract, (err) => {
+                    if (err) {
+                      contract.error = err.message || err;
+                      if (contract.error === constants.blockchain.gasAllowanceError) {
+                        self.logger.error(`[${contract.className}]: ${constants.blockchain.gasAllowanceErrorMessage}`);
+                      } else {
+                        self.logger.error(`[${contract.className}]: ${err.message || err}`);
+                      }
+                      errors.push(err);
+                    }
+                    callback();
+                  });
+                }
```

Comment:
@jrainville as you can see, I'm leveraging the existing `deploys` list instead of introducing another "pipeline" for `async.auto`. The reason being is that, in a way, both scenarios have the same semantics, as in, they perform something that results in our Smart Contract object having an address.

In other words, we just need to check if we have an `addressHandler`, if yes we know we need to compute it (by calling the handler), if not, we know we need to compute it as well (by deploying it).

Hope this makes sense.

Notice however, as discussed in chat, this really only works when users define an `abiDefinition` on the Smart Contract configuration that they want to compute the address for (even if they aren't interested in interacting with it in through RPC clients).

Otherwise, Embark won't consider that Smart Contract for deployment/address determination.

Reply:
Can you explain why it's impossible or too difficult to determine or produce the `abiDefinition` if the user doesn't provide it?  The reason may be quite simple, but as I noted in another comment, contract stuff is not my area of expertise and I'd like to understand better if the need for the user to provide it is a current limitation of Embark or if it's strictly necessary because of how contracts work.

Reply:
> Can you explain why it's impossible or too difficult to determine or produce the abiDefinition if the user doesn't provide it?

That's in the case the contract is not on disk and also not provided by the user as `abiDefinition`. In that case, Embark will just "ignore" the contract and put a warning saying that we don't have the code for that contract.

Anyway, for Richard's issue, there is no problem since we have the contract code.

Reply:
Just adding to what @jrainville for clarity:

Simply said: Embark will only deploy what it has also compiled in the same process (unless an ABIDefinition is provided).

This is not entirely true as you know it also won't deploy when an address is given cause then we know the contract is already deployed.

Reply:
I agree with @michaelsbradleyjr - theoretically if we are determining the address, we know we have an already-deployed contract, and in that case, couldn't we get the ABI from the contract on the network? I believe Embark's current lifecycle would not support this setup as it checks for ABI definition or compilation before considering the contract to be included in the deploy, but I think that it be something to consider as a future feature.

Reply:
> theoretically if we are determining the address, we know we have an already-deployed contract, and in that case, couldn't we get the ABI from the contract on the network?

@emizzle Actually, I didn't know you can fetch the ABI for a contract from the network.

To be very honest, I'm not very happy with the fact that this handler only works if the corresponding contract has been compiled. I've discussed my concerns on Discord with @jrainville as well. Right now, it's really just @richard-ramos use case we're implementing here but it's not generalized so that it works for anybody else who doesn't own the source of the contract in question.

In other words, in my opinion, as soon as there's either an `address` or an `addressHandler`, Embark should be able to process those in the deployment routine, without the requirement of having those contracts compiled before (for `address` this actually applies).

Right now that doesn't work because Embark kicks contract object out that doesn't have either `code` or an `abiDefinition`, while what we're asking for is really just an address, where none of the two properties are needed.